lms/djangoapps/verify_student/management/commands/send_verification_expiry_email.py

```diff
@@ -0,0 +1,166 @@
+            default=10,
+            help='Sleep time in seconds between update of batches')
+
+    def __init__(self, *args, **kwargs):
+        super(Command, self).__init__(*args, **kwargs)
+        self.sspv = SoftwareSecurePhotoVerification.objects.filter(status='approved',
+                                                                   expiry_date__lt=datetime.now(UTC)
+                                                                   ).order_by('user_id')
+        subject = _("Your {platform_name} Verification has Expired").format(
+            platform_name=settings.PLATFORM_NAME
+        )
+        self.verification_expiry_email_vars = {
+            'platform_name': settings.PLATFORM_NAME,
+            'lms_verification_link': '{}{}'.format(settings.LMS_ROOT_URL, reverse("verify_student_reverify")),
+            'help_center_link': settings.ID_VERIFICATION_SUPPORT_LINK
+        }
+        self.email_context = {
+            'subject': subject,
+            'template': 'emails/verification_expiry_email.txt',
+        }
+
+    def handle(self, *args, **options):
+        """
+        Handler for the command
+        """
+        resend_days = options['resend_days']
+        batch_size = options['batch_size']
+        sleep_time = options['sleep_time']
+
+        date_resend_days_ago = datetime.now(UTC) - timedelta(days=resend_days)
+
+        try:
+            max_user_id = self.sspv.last().user_id
+            batch_start = self.sspv.first().user_id
+            batch_stop = batch_start + batch_size
+
+        except AttributeError:
+            logger.info("AttributeError: No approved expired entries found in SoftwareSecurePhotoVerification")
+            return
+
+        while batch_start <= max_user_id:
+            batch_queryset = self.sspv.filter(user_id__gte=batch_start, user_id__lt=batch_stop)
+            users = batch_queryset.values('user_id').distinct()
+
+            for user in users:
+                verification = self.find_recent_verification(user['user_id'])
+                if not verification.expiry_email_date or verification.expiry_email_date < date_resend_days_ago:
+                    user = User.objects.get(id=verification.user_id)
+                    self.verification_expiry_email_vars['full_name'] = user.profile.name
+                    self.email_context['email_vars'] = self.verification_expiry_email_vars
+                    self.email_context['email'] = user.email
+
+                    verification_qs = self.sspv.filter(pk=verification.pk)
+                    send_verification_expiry_email.delay(self.email_context, verification_qs)
+
+            if batch_stop < max_user_id:
+                time.sleep(sleep_time)
+
+            batch_start = batch_stop
+            batch_stop += batch_size
+
+    def find_recent_verification(self, user_id):
+        """
+        Find the most recent expired verification for user.
+        """
+        return self.sspv.filter(user_id=user_id).latest('updated_at')
```

Comment:
Every single invocation of this is going to cause a full table filesort.

```python
>>> from lms.djangoapps.verify_student.models import SoftwareSecurePhotoVerification
>>> from datetime import datetime
>>> from pytz import UTC
>>> sspv = SoftwareSecurePhotoVerification.objects.filter(status='approved', expiry_date__lt=datetime.now(UTC)).order_by('user_id')
>>> find_recent_verification = sspv.filter(user_id=11)
>>> str(sspv.query)
'SELECT `verify_student_softwaresecurephotoverification`.`id`, `verify_student_softwaresecurephotoverification`.`status`, `verify_student_softwaresecurephotoverification`.`status_changed`, `verify_student_softwaresecurephotoverification`.`user_id`, `verify_student_softwaresecurephotoverification`.`name`, `verify_student_softwaresecurephotoverification`.`created_at`, `verify_student_softwaresecurephotoverification`.`updated_at`, `verify_student_softwaresecurephotoverification`.`face_image_url`, `verify_student_softwaresecurephotoverification`.`photo_id_image_url`, `verify_student_softwaresecurephotoverification`.`receipt_id`, `verify_student_softwaresecurephotoverification`.`display`, `verify_student_softwaresecurephotoverification`.`submitted_at`, `verify_student_softwaresecurephotoverification`.`reviewing_user_id`, `verify_student_softwaresecurephotoverification`.`reviewing_service`, `verify_student_softwaresecurephotoverification`.`error_msg`, `verify_student_softwaresecurephotoverification`.`error_code`, `verify_student_softwaresecurephotoverification`.`photo_id_key`, `verify_student_softwaresecurephotoverification`.`copy_id_photo_from_id`, `verify_student_softwaresecurephotoverification`.`expiry_date`, `verify_student_softwaresecurephotoverification`.`expiry_email_date` FROM `verify_student_softwaresecurephotoverification` WHERE (`verify_student_softwaresecurephotoverification`.`status` = approved AND `verify_student_softwaresecurephotoverification`.`expiry_date` < 2019-01-25 19:49:55.439262) ORDER BY `verify_student_softwaresecurephotoverification`.`user_id` ASC'
```

Then the explain for this:

```
mysql> explain SELECT `verify_student_softwaresecurephotoverification`.`id`, `verify_student_softwaresecurephotoverification`.`status`,
    -> `verify_student_softwaresecurephotoverification`.`status_changed`, `verify_student_softwaresecurephotoverification`.`user_id`, `verify_student_softwaresecurephotoverification`.`name`, `verify_student_softwaresecurephotoverification`.`created_at`,
    -> `verify_student_softwaresecurephotoverification`.`updated_at`, `verify_student_softwaresecurephotoverification`.`face_image_url`, `verify_student_softwaresecurephotoverification`.`photo_id_image_url`, `verify_student_softwaresecurephotoverification`.`receipt_id`, `verify_student_softwaresecurephotoverification`.`display`, `verify_student_softwaresecurephotoverification`.`submitted_at`, `verify_student_softwaresecurephotoverification`.`reviewing_user_id`, `verify_student_softwaresecurephotoverification`.`reviewing_service`, `verify_student_softwaresecurephotoverification`.`error_msg`, `verify_student_softwaresecurephotoverification`.`error_code`, `verify_student_softwaresecurephotoverification`.`photo_id_key`, `verify_student_softwaresecurephotoverification`.`copy_id_photo_from_id`, `verify_student_softwaresecurephotoverification`.`expiry_date`, `verify_student_softwaresecurephotoverification`.`expiry_email_date`
    -> FROM `verify_student_softwaresecurephotoverification`
    -> WHERE (`verify_student_softwaresecurephotoverification`.`status` = 'approved' AND `verify_student_softwaresecurephotoverification`.`expiry_date` < '2019-01-25') ORDER BY `verify_student_softwaresecurephotoverification`.`user_id` ASC
    -> \G
*************************** 1. row ***************************
           id: 1
  select_type: SIMPLE
        table: verify_student_softwaresecurephotoverification
         type: range
possible_keys: verify_student_softwaresecu_expiry_date_5c297927
          key: verify_student_softwaresecu_expiry_date_5c297927
      key_len: 9
          ref: NULL
         rows: 1
        Extra: Using index condition; Using where; Using filesort
1 row in set (0.00 sec)
```

Reply:
I think you have a mistake in above query. It should be 
`str(find_recent_verification.query)` instead of `str(sspv.query)`


``` python
>>> sspv = SoftwareSecurePhotoVerification.objects.filter(status='approved', expiry_date__lt=datetime.now(UTC)).order_by('user_id')
>>> find_recent_verification = sspv.filter(user_id=1)
>>> str(find_recent_verification.query)
```

``` mysql
| id | select_type | table                                          | type | possible_keys                                                                                            | key                                                     | key_len | ref   | rows | Extra       |
+----+-------------+------------------------------------------------+------+----------------------------------------------------------------------------------------------------------+---------------------------------------------------------+---------+-------+------+-------------+
|  1 | SIMPLE      | verify_student_softwaresecurephotoverification | ref  | verify_student_softwaresecurephotoverification_fbfc09f1,verify_student_softwaresecu_expiry_date_5c297927 | verify_student_softwaresecurephotoverification_fbfc09f1 | 4       | const |    2 | Using where |
```

Basically I am performing all the filtering on `sspv` (which is filtered queryset and is already ordered). filesort will not be called on it again and again. It will only be called once (as mentioned above because of ordering clause in class meta filesort is inevitable)